### ISSUE TITLE
Facebook's dirty links to external web pages not fully cleaned 

### ISSUE THREAD

**Issue by toughdiamond** [Cimbali/CleanLinks#48]
Facebook seems to have recently taken additional steps to craftily embed tracking data into external links, which CleanLinks 3.2.1 doesn't remove.  For example, here's the link Facebook has for a bbc.com web page before any cleaning at all:
https://l.facebook.com/l.php?u=https%3A%2F%2Fbbc.in%2F2D8dMwt%3Ffbclid%3DIwA.........  [etc. etc., goes on for about 700 characters]
	Obviously a job for CleanLinks, so I rightclick and do "Copy clean link," which gives this:
https://bbc.in/2D8dMwt?fbclid=IwAR1zDT_KSlH90LBrbPFKtdbfVVlRKo_pSPysgVVfzwif1WrILbteg-3rj8s
Clearly a big improvement, well done, but there's still tracking data following the question mark that hasn't been removed.  I can pretty easily delete it manually which gives me this:
https://bbc.in/2D8dMwt
	It'd be good if the next upgrade of CleanLinks would do that itself of course, as the extra data is presumably there to track me.   
	But there's more.  If I paste the above (seemingly clean) link into the address bar and hit go, the link turns into this:
https://www.bbc.com/news/world-latin-america-45982501?ocid=socialflow_facebook&ns_source=facebook&ns_mchannel=social&ns_campaign=bbcnews
	That's clearly not as clean as it looked.  Doesn't seem like much to worry much about in terms of invading privacy, and I stand some chance of cleaning it myself by hitting the Escape key immediately after hitting go, and deleting from the right up to the question mark, to give this, which is the truly clean link I wanted from the start:
https://www.bbc.com/news/world-latin-america-45982501
	After that I can hit go and the page opens fine.  Naturally it's a cumbersome and not very reliable workaround. Like I say, the dirt seems fairly harmless anyway, but what bothers me is that Facebook now seems to have the potential to embed something more worrying than that if it ever decides to, and CleanLinks currently won't stop it.
	By the way, congratulations on creating an otherwise excellent link cleaner.  Best I've found so far :-)

**Comment by Cimbali** [comment]
OK so there's 2 things going on here:

1. There's a `fbclid` which we should remove in the initial link. You can add `|fbclid` at the end of "Remove from Links" in the CleanLinks preferences, and that will do the trick. I'll keep the bug open until I can add this into CleanLinks.

2. The second is that the link you're being redirected to is a link shortener. It's bbc's own link shortener, but in itself it's the same as bit.ly, goo.gl and all those alternate services. The destination ("https://www.bbc.com/news/world-latin-america-45982501") I'm afraid is not available in the link, it's some form of database lookup that bbc.in does, mapping the short link to the long one.

   Turns out it's a specialized link for facebook, so I'm guessing it's auto-generated by some "share to facebook" button, and pressing a "share to twitter button" would generate another `bbc.in` link that will redirect to `https://www.bbc.com/news/world-latin-america-45982501?ocid=socialflow_twitter&ns_source=twitter&...`

   We could clean this at the moment the request is being done, that is bbc.in tells you "HTTP 301 redirect to page `bbc.com... ns_source=facebook`", and when the browser makes that request we remove the offending parameters. However, this requires cleaning parameters from links without having redirects cleaned, so it's blocked by #27 for now.

**Comment by toughdiamond** [comment]
OK,  adding  |fbclid worked fine.  Thanks for the explanation of how the BBC redirects work.  Hope it becomes fixable one day.

**Comment by toughdiamond** [comment]
Today I found another apparent problem with Facebook - I got a private message with a link to a YouTube video in it.  The link superficially looked clean enough: 
https://www.youtube.com/watch?v=-H9G3XmQ03c&feature=share 
but when I did Copy Clean Link, I got this on the clipboard:
https://l.facebook.com/l.php?u=https%3A%2F%2Fwww.youtube.com%2Fwatch%3Fv%3D-H9G3XmQ03c%26feature%3Dshare%26fbclid%3DIwAR2ztGxwZICudzr3UlyqFJCwLz5oEPGdce6touI3P [etc.]
Only seen it with that one private message so far, never on the newsfeed.  A workaround is to select it in the message and do Ctrl-c .

**Comment by Cimbali** [comment]
Now that `|fbclid` is added by default, this bug is blocked by #27 , after which it is only a "whitelist" request to remove `ocid` and `ns_source` from bbc.com URLs.

**Comment by toughdiamond** [comment]
Sorry to post this in what is probably the wrong place, but I'm having great trouble finding the right place to post it, as the whole Github thing is like nothing I've ever seen before and it doesn't seem at all clear how to simply comment on a new version.

**Comment by Cimbali** [comment]
The youtube link not being cleaned seems to be the whitelist acting up again, specifically the `|watch%3Fv` in "Skip links matching with". These rules are too coarse-grain, I'm working on a better system right now, though it might take some time.

I've created a new issue for the missing clean links menu item, as removing it was not intended. In the mean time, any URLs which you feel should have been cleaned but aren't you can report here, it'll help build a better set of rules.

**Comment by toughdiamond** [comment]
OK, of course I'm back to using v.3.2.1 for this because the update doesn't work for me.   Here's another example which I got today in Messenger when I rightclicked it and did Copy Clean Link:
https://l.facebook.com/l.php?u=https%3A%2F%2Fwww.youtube.com%2Fwatch%3Fv%3Dfre89067f1g%26feature%3Dshare%26fbclid%3DIwAR3mbO3guzb47dldxkrbUdCLMnaaNbQUGPWmz_UO2LMdaAXbMHIC994UaC8&h=AT0mTl5x2_Se3UiPRtdmWbwznQgVXOCcrtc35RoHejcWxFa0MjjpzIIGOjsdI9OvEGUTD_kT__1dfFWvDQZHusEonyvnjY6-8yZfL2EgXjgjczpPx8fO1Gyg-U7u6uNnFPnm

Of course the clean link is https://www.youtube.com/watch?v=fre89067f1g
Seems to happen every time I get sent a link on Messenger, as if Clean Links is doing nothing, though it's usually fine for Newsfeed links.
Incidentally, are you OK about supporting 32-bit Windows 7 anyway?  As I've said previously, that's what I'm using, and I appreciate that you might have better things to do than cater for an OS that's rapidly going out of fashion.

**Comment by Cimbali** [comment]
For the youtube links you should remove the `|watch%3Fv` that I mentioned from the "Skip Links matching with" preference. It might break something else, but it's rather unlikely.

For the OS, there should be no difference in theory. Only the browser's version should have an impact because it provides all the interface we use -- we never interact with the OS directly.

**Comment by toughdiamond** [comment]
Thanks, removing that bit of script did the trick.
